### PR TITLE
Implement ES2021 WeakRef and FinalizationRegistry

### PR DESCRIPTION
**Pull request by anivar**
Implements ES2021 WeakRef and FinalizationRegistry following the NativeWeakMap pattern.

## Implementation

Two self-contained classes, about 400 lines total:
- NativeWeakRef.java - uses Java's WeakReference directly
- NativeFinalizationRegistry.java - uses PhantomReference + ReferenceQueue for GC detection

The finalization approach is simple: poll the ReferenceQueue during register() calls to process any pending cleanups. No separate queue management or Context modifications needed.

## Android Compatibility

Uses PhantomReference instead of the Cleaner API since Cleaner isn't available on Android. This keeps the implementation portable across all Java 8+ platforms.

## ES2021 Standard Compliance

Correctly implements symbol handling per spec:
- Allows unregistered symbols as weak targets
- Rejects registered symbols created with Symbol.for since they persist globally
- Validates all target/heldValue/token constraints
- Proper Symbol.toStringTag on prototypes

## Test Results

All 76 test262 tests passing:
- 29 WeakRef tests
- 47 FinalizationRegistry tests

Also fixes 2 Object.seal tests.

Thread-safe implementation using ConcurrentHashMap for registration tracking.

### CODE REVIEW

File: rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+    private static final String REGISTER_METHOD = "register";
+    private static final String UNREGISTER_METHOD = "unregister";
+    private static final String CLEANUP_SOME_METHOD = "cleanupSome";
+
+    // Constants
+    private static final String MSG_NO_CALLBACK = "msg.finalization.registry.no.callback";
+    private static final String MSG_REGISTER_NOT_OBJECT =
+            "msg.finalization.registry.register.not.object";
+    private static final String MSG_SAME_TARGET = "msg.finalization.registry.register.same.target";
+    private static final int CONSTRUCTOR_ARITY = 1;
```
Comment (gbrail) [inline rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java]: I don't know if I normally care too much, but in general in this project we have NOT been using constants for every arity and method name parameter. I know that's not what they taught me in school, but for this project having all that information together in the one line of code where we use it makes more sense to me than all the constant back and forth.

File: rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+                        NativeFinalizationRegistry::constructor);
+        constructor.setPrototypePropertyAttributes(DONTENUM | READONLY | PERMANENT);
+        return constructor;
+    }
+
+    private static void configurePrototype(LambdaConstructor constructor, Scriptable scope) {
+        // Define prototype methods
+        constructor.definePrototypeMethod(
+                scope,
+                REGISTER_METHOD,
```
Comment (gbrail) [inline rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java]: See above -- I don't think that we need constants for method names and arities.

File: rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+        if (args.length < 2) {
+            throw ScriptRuntime.typeErrorById(MSG_REGISTER_NOT_OBJECT);
+        }
+
+        Object target = args[0];
+        Object heldValue = args[1];
+        Object unregisterToken = args.length > 2 ? args[2] : null;
+
+        if (!isValidTarget(target)) {
+            throw ScriptRuntime.typeErrorById(MSG_REGISTER_NOT_OBJECT);
```
Comment (gbrail) [inline rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java]: I don't think that it adds to clarity to have separate constants for message Ids either, and the rest of the code in the project just puts the message name in there, which saves an indirection when trying to find it in my opinion!

File: rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+     * @param heldValue the held value to pass to the callback
+     */
+    private void executeCleanupCallback(Object heldValue) {
+        if (cleanupCallback == null) {
+            return;
+        }
+
+        Context cx = Context.getCurrentContext();
+        if (cx == null) {
+            cx = Context.enter();
```
Comment (gbrail) [inline rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java]: We have a long thread going on about this elsewhere, but at least for now you can and should use try-with-resources here rather than exiting the context in a finally clause.

File: rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+        try {
+            Scriptable scope = callback.getParentScope();
+            if (scope == null) {
+                scope = cx.initStandardObjects();
+            }
+            callback.call(cx, scope, scope, new Object[] {heldValue});
+        } catch (RhinoException e) {
+            // Cleanup callback errors should not break the cleanup process
+            // In a production environment, this might be logged
+        } catch (Exception e) {
```
Comment (gbrail) [inline rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java]: Is there a reason why we'd legitimately want to catch a non-JavaScript exception here? Otherwise we should just catch RhinoException, assuming that the spec says that we still finalize even if the callback throws.

File: rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java [hunk 6]
```diff
@@ -0,0 +1,462 @@
...
+    private void callCleanupCallback(Context cx, Object heldValue, Function callback) {
+        try {
+            Scriptable scope = callback.getParentScope();
+            if (scope == null) {
+                scope = cx.initStandardObjects();
+            }
+            callback.call(cx, scope, scope, new Object[] {heldValue});
+        } catch (RhinoException e) {
+            // Cleanup callback errors should not break the cleanup process
+            // In a production environment, this might be logged
```
Comment (gbrail) [inline rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java]: Not to be annoying, but this code will absolutely be used in a production environment ;-)

File: rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+     * Calls the specified cleanup callback with the held value.
+     *
+     * @param cx the context to use
+     * @param heldValue the held value to pass to the callback
+     * @param callback the callback function to call
+     */
+    private void callCleanupCallback(Context cx, Object heldValue, Function callback) {
+        try {
+            Scriptable scope = callback.getParentScope();
+            if (scope == null) {
```
Comment (gbrail) [inline rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java]: Is there a reason why the scope would be null here?

File: tests/testsrc/test262.properties [hunk 8]
```diff
@@ -948,7 +948,21 @@ built-ins/Error 7/53 (13.21%)
     cause_abrupt.js
     proto-from-ctor-realm.js
 
-~built-ins/FinalizationRegistry
```
Comment (gbrail) [inline tests/testsrc/test262.properties:943]: This is very nice progress -- do you know why the remaining tests are failing?

File: rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java [hunk 9]
```diff
@@ -0,0 +1,462 @@
...
+            processCleanupWithCallback(ref, callbackToUse, cx);
+        }
+    }
+
+    /**
+     * Processes cleanup for a specific weak reference.
+     *
+     * @param ref the weak reference to process
+     */
+    private void processCleanup(FinalizationWeakReference ref) {
```
Comment (gbrail) [inline rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java]: Looking at the code coverage, it looks like we don't have test coverage for this and the next five functions (processCleanupWithCallback, executeCleanupCallback, callCleanupCallback). Not sure how you can test those, but even some "white box tests" might help just to ensure that we have tests for it.

**Review by gbrail — COMMENTED** [review]
Thanks -- this looks like good progress. I have a bunch of questions, please take a look -- thanks!

File: rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java (hunk 1 above)
Comment (anivar) [reply rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java]: You're absolutely right about the project style. Looking at similar classes like NativeWeakMap, I can see that method names and arities are typically inlined directly in
  the definePrototypeMethod calls rather than using constants. I'll remove these constants:
  - REGISTER_METHOD → inline "register"
  - UNREGISTER_METHOD → inline "unregister"
  - CONSTRUCTOR_ARITY → inline 1
  - REGISTER_ARITY → inline 2
  - UNREGISTER_ARITY → inline 1

  This will make the code more consistent with the rest of the codebase and reduce the indirection as you mentioned.

File: rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java (hunk 2 above)
Comment (anivar) [reply rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java]: I'll replace these constants with inline values in the definePrototypeMethod calls

File: rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java (hunk 3 above)
Comment (anivar) [reply rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java]: I will follow  the project's pattern of putting the message ID directly in the error call, which as you noted, makes it easier to find the message when debugging.

File: rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java (hunk 4 above)
Comment (anivar) [reply rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java]: Will do 
```

try (Context enteredCx = Context.enter()) {
      callCleanupCallback(enteredCx, heldValue);
  }
```

File: rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java (hunk 5 above)
Comment (anivar) [reply rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java]: you're right that we should only catch RhinoException here

File: rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java (hunk 6 above)
Comment (anivar) [reply rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java]: Will try  

```
 } catch (RhinoException e) {
 
      Context.reportWarning("FinalizationRegistry cleanup callback threw: " + e.getMessage());
  }
```

File: rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java (hunk 7 above)
Comment (anivar) [reply rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java]: A JavaScript function passed to new FinalizationRegistry(callback) will always have a parent scope  which was defined somewhere.

  If somehow it's null, that's a Rhino bug we shouldn't hide. Creating a new scope with initStandardObjects() would just mask the problem and cause weird behavior.

  I'll remove the null check and trust the scope exists

File: tests/testsrc/test262.properties (hunk 8 above)
Comment (anivar) [reply tests/testsrc/test262.properties:943]: The remaining failures fall into a few categories:

  1. cleanupSome (14 tests) - This optional method isn't implemented yet. It allows synchronous triggering of cleanup callbacks, which is complex to implement safely.
  2. Cross-realm tests - proto-from-ctor-realm.js fails because of how Rhino handles cross-realm prototype chains differently from the spec.
  3. Edge cases in register/unregister - Some tests like custom-this.js and holdings-any-value-type.js are failing due to subtle differences in type checking or parameter handling.
  4. Weak reference constraints - Tests like throws-when-target-cannot-be-held-weakly.js will be addressed when the WeakRef PR lands, as that will establish the proper constraints for what can be held weakly.

  The core functionality works - objects can be registered, cleanup callbacks fire when GC happens, and unregister works. The failures are mostly edge cases and the optional cleanupSome feature.

File: rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java (hunk 9 above)
Comment (anivar) [reply rhino/src/main/java/org/mozilla/javascript/NativeFinalizationRegistry.java]: The cleanup methods (processCleanup, executeCleanupCallback, etc.) only run when GC actually collects registered objects.

  For test coverage, I could add white box tests that use reflection to test these methods directly:

```
  @Test
  public void testProcessCleanupDirectly() throws Exception {
      Method processCleanup = NativeFinalizationRegistry.class
          .getDeclaredMethod("processCleanup", FinalizationWeakReference.class);
      processCleanup.setAccessible(true);
      // ... test the method
  }
```

**Review by aardvark179 — CHANGES_REQUESTED** [review]
I thought I'd written a review of this, but I must have closed the tab or something.

So, according to the spec the cleanup routines should be run when the weak ref target is no `EMPTY`, so this should probably be written using `PhantomReference`s (they are put on the queue when their referent has been collected). I am also unsure what the effect is on the JVM of create a large number of processor queues, and they can be tricky to manage correctly.

Could you try implementing this using a single queue (maybe using a `java.lang.ref.Cleaner` and putting the tasks from there onto a queue in each `Context`. We could likely combine processing of those with the micro task loop run as the context is exited.

I think that would help avoid large reference queues hanging around, and would make it easier to understand when things can be run.

**Comment by anivar** [comment]
@aardvark179 Thank you for the detailed review! Yes, he did leave inline comments earlier and I responded to them and followed the suggestions. However, after refactoring the implementation, the filenames changed and I can't access those same comments anymore.

I understand your architectural suggestions:
1. Using PhantomReferences instead of WeakReferences for more deterministic cleanup behavior
2. Implementing a single shared ReferenceQueue across all FinalizationRegistry instances for efficiency
3. Integrating the cleanup process with the Context's micro-task loop for better timing control

**Comment by anivar** [comment]
@aardvark179 Implementation completely rewritten per your feedback:

PhantomReferences now replace WeakReferences for correct GC semantics. Added shared ReferenceQueue with Context integration for cleanup processing in the JavaScript execution loop as requested.

Key improvements:
- Thread-safe operations with ConcurrentHashMap
- O(1) token-based unregistration 
- Memory leak prevention with proper cleanup
- Significant code simplification 
- All deprecation and ErrorProne warnings properly resolved
- 72+ tests passing including cross-realm scenarios


Ready for re-review!

**Comment by aardvark179** [comment]
Sorry I haven’t got to re-reviewing this one yet. It’s been a busy few weeks, but I promise I will get to it.

It would also be great to have a chat at some point so I can understand your Rhino use case better and maybe help plan out which features are the best ones to tackle first to enable that.

**Comment by anivar** [comment]
No worries @aardvark179! 

I'd be interested in that chat about Rhino use cases and feature prioritization. Feel free to reach out whenever works for your schedule.

Thanks for the architectural guidance on PhantomReference and shared queue - it made the implementation much cleaner.

**Comment by anivar** [comment]
Force pushed a simplified implementation that follows the NativeWeakMap pattern. All tests passing, Android-compatible, no Context modifications needed. Ready for review.
